libs/langgraph/src/langgraph-architecture-spec.ts

```diff
@@ -0,0 +1,399 @@
+/**
+ * LangGraphJS Architecture Specification
+ *
+ * This document outlines the key architectural components and their relationships
+ * within the LangGraphJS library.
+ */
+
+/**
+ * ===========================
+ * Core Architecture Overview
+ * ===========================
+ *
+ * LangGraphJS implements a message-passing computation model based on Google's Pregel
+ * system. The architecture has four main layers with clear dependencies:
+ *
+ * 1. Channels Layer - Foundation for state management and communication
+ * 2. Checkpointer Layer - Persistence and serialization of system state
+ * 3. Pregel Layer - Core execution engine implementing the message-passing model
+ * 4. Graph Layer - High-level interfaces for building workflows
+ */
+
+/**
+ * =========================
+ * Pregel System Overview
+ * =========================
+ *
+ * The Pregel system implements a bulk synchronous parallel (BSP) processing model:
+ *
+ * - Supersteps: Discrete execution phases
+ * - Vertices (Nodes): Processing units that read from and write to channels
+ * - Channels: Communication mediators that manage state and message passing
+ * - Edges: Defined by subscription relationships between nodes and channels
+ * - Checkpoints: Persistent snapshots of system state
+ *
+ * Communication Model:
+ * - Nodes NEVER communicate directly with each other
+ * - All communication is mediated through Channels
+ * - Nodes subscribe to Channels to receive data
+ * - Nodes write to Channels to share data
+ * - Channels can customize how values are accumulated (last value, append to list, etc.)
```

Comment:
Worth mentioning values here? it's kind of mentioned in the last point, but not explicitly